### PR TITLE
fix: conditional binding (#23)

### PR DESCRIPTION
**Pull request by mmaietta**
Enhancement and fixes: #23 (Module did not self-register)
#23 occurs because of Linux having no export defined due to having no sources
https://github.com/codebytere/node-mac-permissions/blob/c74d338c4e86d85181e8f51b50624f5e61024889/binding.gyp#L4-L10

Adding conditional binding so that we can retain type-safety of the import for things like `permissions.AuthType` as function params with a generic import that is platform-agnostic. 
```
import * as permissions from 'node-mac-permissions'
async requestSystemAccessPermission (accessType: permissions.AuthType): Promise<permissions.PermissionType> { }
```

---

Adding an additional test to run on a separate Linux node for this use case

Opening as DRAFT because I'm unable to actually test this on Linux, only verified through Github actions Ubuntu node and through headless docker container (via `xvfb-run`). Going to test on windows soon. Wanted to get a head start on any feedback though to make sure I'm not already off the rails

### CODE REVIEW

**Comment by mmaietta** [comment]
@codebytere it seems that xcode 12.4 changed some things and now it throws this error when attempting to build:
```
warning: /Applications/Xcode_12.4.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: archive library: Release/nothing.a the table of contents is empty (no object file members in the library define global symbols)
  CXX(target) Release/obj.target/permissions/permissions.o
../permissions.mm:274:8: error: duplicate case value: 'kCLAuthorizationStatusAuthorized' and 'kCLAuthorizationStatusAuthorizedAlways' both equal '3'
  case kCLAuthorizationStatusAuthorizedAlways:
       ^
../permissions.mm:273:8: note: previous case defined here
  case kCLAuthorizationStatusAuthorized:
       ^
1 error generated.
make: *** [Release/obj.target/permissions/permissions.o] Error 1
```

I've commented out that line since they're duplicate values and left an additional comment on the new behavior.

**Comment by codebytere** [comment]
@mmaietta if you rebase it should be good now

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -10,14 +10,17 @@ on:
...
+      matrix:
+        os: [macos-latest, ubuntu-latest]
     steps:
     - uses: actions/checkout@master
     - name: Use Node.js 14.x
       uses: actions/setup-node@v1
       with:
-        version: 14.x
-    - name: npm install, build, and test
+        node-version: 14.x
```
Comment (codebytere) [inline .github/workflows/test.yml:22]: ooh thanks! been meaning to get to this

File: index.js [hunk 2]
```diff
@@ -1,4 +1,21 @@
...
+  askForMicrophoneAccess: nonMacResponse,
+  askForPhotosAccess: nonMacResponse,
+  askForSpeechRecognitionAccess: nonMacResponse,
+  askForScreenCaptureAccess: nonMacResponse,
+  askForAccessibilityAccess: nonMacResponse,
+  getAuthStatus: nonMacResponse,
+}
+
+const os = require('os')
+const permissions = os.platform() === 'darwin' ? require('bindings')('permissions.node') : stub
```
Comment (codebytere) [inline index.js:18]: nit but let's use either `os.platform()` or `process.platform` everywhere?

File: test/module.spec.js [hunk 3]
```diff
@@ -1,18 +1,19 @@
 const { expect } = require('chai')
-const {
-  askForFoldersAccess,
-  getAuthStatus,
-} = require('../index')
+const permissions = require('../index')
+
+const isMac = process.platform === 'darwin'
```
Comment (codebytere) [inline test/module.spec.js:4]: see above comment

File: index.js [hunk 4]
```diff
@@ -34,16 +51,7 @@ function askForFoldersAccess(folder) {
 }
 
 module.exports = {
-  askForCalendarAccess: permissions.askForCalendarAccess,
-  askForContactsAccess: permissions.askForContactsAccess,
+  ...permissions,
```
Comment (codebytere) [inline index.js:54]: we probably also need to handle calls to `askForFoldersAccess` and `getAuthStatus` - `permissions.askForFoldersAccess.call(this, folder)` @ L33 for example probably blows up, does it not?

File: package.json [hunk 5]
```diff
@@ -7,8 +7,8 @@
   "scripts": {
     "build": "node-gyp rebuild",
     "clean": "node-gyp clean",
-    "lint": "prettier --check index.js",
-    "format": "clang-format -i permissions.mm && prettier --write index.js",
+    "lint": "prettier --check index.js test/module.spec.js",
+    "format": "clang-format -i permissions.mm && prettier --write index.js test/module.spec.js",
```
Comment (codebytere) [inline package.json:11]: ty for this too

File: index.js (hunk 2 above)
Comment (mmaietta) [reply index.js:18]: I have a reason for the difference here. I actually am not sure if `process.platform` could be considered "dead code" by a bundler at compile time in the same manner that `process.env.NODE_ENV === 'development'` could be stripped from a prod bundle.

I was keeping the require's in the tests to a minimum while using a runtime eval for the conditional binding. I can update the test to use `os.platform()`

File: index.js (hunk 4 above)
Comment (mmaietta) [reply index.js:54]: These should also be covered in the additional unit test that runs on Linux 🙂 

https://github.com/codebytere/node-mac-permissions/blob/321a1f80896fa517983a79aaeffbd2d0af6e3f62/test/module.spec.js#L67-L68

The type/folder is validated, and the call is still to `permissions.askForFoldersAccess: () => undefined` from the original stub, so it behaves as expected with the test's assert. I originally had `nonMacResponse = (...args) => undefined` but then `...args` was just flagged as unused.

These two tests also run on all nodes to enforce consistency of the arg cross-platform.
https://github.com/codebytere/node-mac-permissions/blob/321a1f80896fa517983a79aaeffbd2d0af6e3f62/test/module.spec.js#L10-L14
https://github.com/codebytere/node-mac-permissions/blob/321a1f80896fa517983a79aaeffbd2d0af6e3f62/test/module.spec.js#L40-L44
The only notable difference is that the `index.d.ts` does not specify the new `undefined` behavior.  I felt that was acceptable as this is explicitly a mac-only module, this enhancement is mainly to allow type-safety with a standard import in local projects.

**Comment by mmaietta** [comment]
Marking as ready for review.

My team and I have already been leveraging this implementation (via patch-package) on our internal dist builds for a couple of months, so I have high confidence in this PR. 
We're running our full test suite for an external prod release soon.

**Pull request closed without merging**